### PR TITLE
Make Cabal-2.2 behave like Cabal-2.0 related to #5055

### PR DESCRIPTION
**Pull request by phadej**
Please include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/#conventions).
* [x] Any changes that could be relevant to users have been recorded in the changelog.
* [x] The documentation has been updated, if necessary.
* [x] If the change is docs-only, `[ci skip]` is used to avoid triggering the build bots.

Please also shortly describe how you tested your change. Bonus points for added tests!

### CODE REVIEW

**Comment by phadej** [comment]
@23Skidoo ready for the review. This is quick'n'dirty backport of Cabal-2.0 functionality.

File: Cabal/tests/ParserTests/regressions/issue-5055.format [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+        base >=4.8 && <5
+
+test-suite flag-cabal-test
+    type: exitcode-stdio-1.0
+    main-is: SecondMain.hs
+    default-language: Haskell2010
+    build-depends:
+        base >=4.8 && <5
+    
+    if os(windows)
```
Comment (23Skidoo) [inline Cabal/tests/ParserTests/regressions/issue-5055.format]: A pretty printer bug?

File: Cabal/Distribution/PackageDescription/Parsec.hs [hunk 2]
```diff
@@ -306,6 +307,16 @@ goSections specVer = traverse_ process
             commonStanzas <- use stateCommonStanzas
             name' <- parseUnqualComponentName pos args
             flib  <- lift $ parseCondTree' (foreignLibFieldGrammar name')  commonStanzas fields
+
+            let hasType ts = foreignLibType ts /= foreignLibType mempty
```
Comment (23Skidoo) [inline Cabal/Distribution/PackageDescription/Parsec.hs]: > `foreignLibType mempty`

Can we write `mempty :: ForeignLibType` or `ForeignLibTypeUnknown` instead? Likewise for test suites and benchmarks.

File: Cabal/Distribution/PackageDescription/Parsec.hs [hunk 3]
```diff
@@ -599,6 +630,28 @@ mergeCommonStanza (CondNode bi _ bis) (CondNode x _ cs) =
     -- tree components are appended together.
     cs' = map (fmap fromBuildInfo) bis ++ cs
 
+-------------------------------------------------------------------------------
+-- Branches
+-------------------------------------------------------------------------------
+
+-- Check that a property holds on all branches of a condition tree
```
Comment (23Skidoo) [inline Cabal/Distribution/PackageDescription/Parsec.hs]: Might be worth mentioning that this function is copy-pasted from `D.PD.Parse`.

File: Cabal/Distribution/PackageDescription/Parsec.hs (hunk 2 above)
Comment (phadej) [reply Cabal/Distribution/PackageDescription/Parsec.hs]: I'd like to refactor that soon after the branch is cut...

File: Cabal/tests/ParserTests/regressions/issue-5055.format (hunk 1 above)
Comment (phadej) [reply Cabal/tests/ParserTests/regressions/issue-5055.format]: ... as this is "bug-compatible" behaviour, with `cabal-install-2.0`. `cabal-2.0` formats the input as

```
name: issue
version: 5055
cabal-version: >=2.0
build-type: Simple
license: BSD3
synopsis: no type in all branches
description:
    no type in all branches.
category: Test

executable  flag-test-exe
    main-is: FirstMain.hs

test-suite  flag-cabal-test
    type: exitcode-stdio-1.0
    main-is: SecondMain.hs
    build-depends:
        base >=4.8 && <5
    default-language: Haskell2010
```

There is no conditional section at all

File: Cabal/Distribution/PackageDescription/Parsec.hs (hunk 2 above)
Comment (23Skidoo) [reply Cabal/Distribution/PackageDescription/Parsec.hs]: Fine.

File: Cabal/tests/ParserTests/regressions/issue-5055.format (hunk 1 above)
Comment (23Skidoo) [reply Cabal/tests/ParserTests/regressions/issue-5055.format]: OK, fine, let's merge as is.

**Comment by 23Skidoo** [comment]
Merged, thanks!

**Pull request merged**